### DISCUSSION TITLE
`Partial` derive macro implementation

### DISCUSSION THREAD

**Discussion by arendjr** [biomejs/biome#1609] [RFC]
I’m starting work on the `Partial` derive macro, which is supposed to take a struct and generate a companion struct where every field is wrapped in `Option` (for context, see my original proposal: https://github.com/biomejs/biome/pull/1460#issuecomment-1883942619).

Now I’m wondering how opinionated I should make that macro. Given that we are introducing this macro for the purpose of easing the way we implement configuration merging, I think somewhat opinionated makes sense.

For instance, all the structs we apply this too also derive serde’s `Serialize` and `Deserialize`. Given that every field will become an `Option`, I think it only makes sense if the macro inserts the line `#[serde(skip_serializing_if = "Option::is_none")]` for you. It doesn’t only save us from repetition, it also prevents mistakes, since I already see a few fields where it was forgotten (I don’t think these omissions were intentional).

But what about `bpaf`? `biome_deserialize` doesn’t depend on `bpaf`, so it feels weird if the macro does similar things there, even though it could. But I think it might be useful if it could just insert `#[bpaf(hide)]` for any fields that don't any other `bpaf` attributes. 

Let me know what you think!

**Comment by ematipico** [comment]
We shouldn't add `bpaf` in the macro.

When I added `bpaf`, documentation was good, but there were still a few hiccups.

Not all the fields of the confirmation should be added to the CLI, and `#[bpaf(hide)]` doesn't prevent the macro from adding bpaf to the children's fields. I believe it's `#[bpaf(pure)]`.

Anyway, `serde` makes sense because we must serialise all the fields. Bpaf, not.

**Reply by pacak** [reply]
`#[bpaf(hide)]` removes an item from help/autocomplete but if you are parsing it - some value is needed. With `#[bpaf(pure)]` or `pure_with` you can specify which value to use instead of trying to parse it.

**Comment by Conaclos** [comment]
Maybe it is time to create `biome_configuration`and `biome_configuration_macros` crates?

**Reply by ematipico** [reply]
Not a bad idea

**Reply by arendjr** [reply]
Not a bad idea indeed! I don’t think I need it in the PR I’m currently preparing, but it appears there may be more opportunities for cleanup opening up, especially around config/settings conversions. Those would very much belong there, I imagine. Not sure when I’ll get to that though 😅

**Comment by Conaclos** [comment]
Implemented.
